[CK_TILE] Support for elementwise kernel

amd-yashagar opened the pull request:
## Proposed changes

Implementation for the elementwise kernel that supports arbitrary number of input tensors for arbitrary dimensions.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

Commit:
Elementwise kernel implementation

Co-authored-by: Sami Aario <samaario@amd.com>
Co-authored-by: Mohsen Saffari <mohsen.saffari@amd.com>
Co-authored-by: yashagar <yashagar@amd.com>

amd-yashagar committed:
Elementwise with generalized nDims

Commit:
Adding the n-ary input tensor feature

amd-yashagar committed:
Generalize dimensions on top of inputs

Commit:
Add TFLOPS + remove std usage for tuples

msaffari-amd committed:
1D basecase optimization

Commit:
Cleanup code + refactoring to a common interface

SamiAario-AMD committed:
Generalize to unary and add an example

Commit:
Cleanup, refactoring and commenting

AviralGoelAMD committed:
Suggestions for LWPCK-3170: elementwise kernel improvements

AviralGoelAMD commented:
Hey team, I made a few suggestions in another PR (feel free to keep/discard). 
https://github.com/ROCm/composable_kernel/pull/2248

amd-yashagar committed:
Clang-format: remod.py

SamiAario-AMD committed:
Replace InputTensorType with XDataType as the type of input_tensors

SamiAario-AMD committed:
Add Tuple::apply and use it in ElementWiseKernel::operator to call operation with the exact number of arguments in xs

aosewski commented on example/ck_tile/37_elementwise/CMakeLists.txt:
Can you please change the example number? To 19 ? (At this moment last is 18) I don't understand why sb started with thirty sth...

SamiAario-AMD replied:
Fixed in commit 4145fbfd.

aosewski commented on example/ck_tile/37_elementwise/elementwise_example.cpp:
Please add copyright header here and in other files.

SamiAario-AMD replied:
Fixed in commit bcccb723.

aosewski commented on example/ck_tile/37_elementwise/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+# Elementwise example targets 2D inputs
+set(TARGET_NAME_2D_INPUT tile_example_elementwise)
+add_executable(${TARGET_NAME_2D_INPUT} elementwise_example.cpp)
+rocm_install(TARGETS ${TARGET_NAME_2D_INPUT} COMPONENT examples)
```
I'm not sure we need this at the moment. @illsilin  Could you please comment on this?

SamiAario-AMD replied:
`rocm_install` lines removed from our `CMakeLists.txt` in commit 96d1a6be. I went ahead with this since none of the other examples have a corresponding line.

aosewski commented on example/ck_tile/37_elementwise/elementwise_example.cpp:
```diff
@@ -0,0 +1,200 @@
+    int warmup            = arg_parser.get_int("warmup");
+    int repeat            = arg_parser.get_int("repeat");
+
+    assert(stride >= N);
```
The `assert` function is a no-op (disabled) in a release build mode. If you want to validate it, throw an error if has incorrect value.

SamiAario-AMD replied:
Fixed in commit 05bbba7b.

aosewski commented on example/ck_tile/37_elementwise/elementwise_example.cpp:
```diff
@@ -0,0 +1,200 @@
+    // Copy data from host tensors to device buffers.
+    x_buf_a.ToDevice(x_host_a.data());
+    x_buf_b.ToDevice(x_host_b.data());
+    y_buf.ToDevice(y_host.data());
```
You'd rather want to set this to zero (if it's needed, or just for sanity). Right now you copy from uninitialized (or in best scenario default initialized) memory region.

SamiAario-AMD replied:
Addressed in commit 64a7d3fc where we no longer copy `y_buf` to the device. Please advise if this is not what you meant.

aosewski commented on example/ck_tile/37_elementwise/elementwise_example.cpp:
```diff
@@ -0,0 +1,200 @@
+    // 3. Configure the kernel execution parameters.
+    // Dividing the problem into blocktile, warptile, and vector
+    // The blocktile is the size of the tile that will be processed by a single thread block (also
+    // called work group) The warptile is the size of the tile that will be processed by a single
```
"Work group" is a HIP (OpenCL) nomenclature. "Thread block" is a CUDA terminology. We should stick to HIP as we're writing in HIP ;)
Please stick to HIP nomenclature throughout the file. You may add CUDA's corresponding names in brackets as well.

SamiAario-AMD replied:
Addressed in commit 821019d7.

aosewski commented on example/ck_tile/37_elementwise/elementwise_example.cpp:
```diff
@@ -0,0 +1,200 @@
+
+    // 4. Create the kernel
+
+    // ElementWiseTraits bundles these tiling parameters.
```
```suggestion
    // ElementWiseShape bundles these tiling parameters.
```

SamiAario-AMD replied:
Addressed in commit 821019d7.

aosewski commented on example/ck_tile/37_elementwise/elementwise_example.cpp:
```diff
@@ -0,0 +1,200 @@
+    // This is often a multiple of the warp size, 64 on CDNA.
+    // Here, it's explicitly set to 512. This should be consistent with Shape::kBlockSize.
+    // Shape::kBlockSize would be BlockWarps * warpSize (e.g., 8 * 64 = 512).
+    constexpr ck_tile::index_t kBlockSize = 64 * BlockWarps::at(ck_tile::number<0>{});
```
We have function for warp size: https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/core/arch/arch.hpp#L51

SamiAario-AMD replied:
Addressed in commit 656557e7.

aosewski commented on example/ck_tile/37_elementwise/elementwise_example_unary.cpp:
```diff
@@ -0,0 +1,131 @@
+    ck_tile::DeviceMem x_buf_a(x_host_a.get_element_space_size_in_bytes());
+    ck_tile::DeviceMem y_buf(y_host.get_element_space_size_in_bytes());
+    x_buf_a.ToDevice(x_host_a.data());
+    y_buf.ToDevice(y_host.data());
```
You don't need to send this to GPU.

SamiAario-AMD replied:
Addressed in commit 64a7d3fc.

aosewski commented on example/ck_tile/37_elementwise/reference_add.hpp:
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Please update copyright headers in all files.
```suggestion
// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```

SamiAario-AMD replied:
Fixed in commit bcccb723.

aosewski commented on example/ck_tile/37_elementwise/reference_add.hpp:
We have reference operations defined here: https://github.com/ROCm/composable_kernel/tree/develop/include/ck_tile/host/reference - they usually run with multiple threads or we have naive GPU kernels to speed up.

damien-lejeune replied:
Fixed in [1046ff91](https://github.com/ROCm/composable_kernel/pull/2246/commits/1046ff914d22bfe4dc4b6aaf8064f13d0f31a663)

aosewski commented on example/ck_tile/37_elementwise/reference_square.hpp:
As above.

damien-lejeune replied:
Fixed in [1046ff91](https://github.com/ROCm/composable_kernel/pull/2246/commits/1046ff914d22bfe4dc4b6aaf8064f13d0f31a663)

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
```suggestion
// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```

SamiAario-AMD replied:
Fixed in commit bcccb723.

aosewski replied:
I don't see this fixed.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,142 @@
+
+namespace ck_tile {
+
+template <typename F, typename Tuple, ck_tile::index_t... Is>
+
+constexpr void for_each_in_tuple_impl(F&& f, Tuple&& t, sequence<Is...>)
+{
+    (f(std::forward<Tuple>(t).get(number<Is>{})), ...);
+}
+
+template <typename F, typename Tuple>
+constexpr void for_each_in_tuple(F&& f, Tuple&& t)
+{
+    for_each_in_tuple_impl(std::forward<F>(f),
+                           std::forward<Tuple>(t),
+                           make_index_sequence<std::tuple_size<std::decay_t<Tuple>>::value>{});
+}
```
Isn't this https://github.com/ROCm/composable_kernel/blob/develop/include/ck/utility/tuple_helper.hpp#L98 what you're looking for ?

aosewski replied:
You might want to move this to `tuple_helper.hpp`

SamiAario-AMD replied:
This code was removed in commit 876b057f.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,142 @@
+                [&](auto idx) {
+                    auto tensor_view = make_naive_tensor_view<address_space_enum::global,
+                                                              memory_operation_enum::set,
+                                                              amd_buffer_coherence_enum::slc>(
```
This would be valid only for arch < gfx94

SamiAario-AMD replied:
We switched to default template parameters for `make_naive_tensor_view` in commit 86519ac8, which should address this.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,142 @@
+
+        const auto x_tiles = load_tiles(x_windows);
+
+        auto y_tile = load_tile(y_window);
```
We do not want to read the output. If the user would like to do `y += x` he should use this kernel this way `y = y + x` - so pass output as one of the inputs. This especially would not make any sense for permutation operation which should be doable by this kernel.

SamiAario-AMD replied:
I believe this commit properly addresses this issue: 2c469077

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,142 @@
+
+    template <typename... XDataType, typename Dims>
+    CK_TILE_DEVICE void operator()(Dims lens,
+                                   Dims strides,
```
Input and output could have different strides (layouts).

SamiAario-AMD replied:
Fixed by @damien-lejeune in commit b52400cd.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,142 @@
+
+            auto y = y_tile(tile_idx);
+
+            apply_operation(ElementWiseOperation{}, y, x_values);
```
Although this approach seem to be nice generic impl for associative operations - we do not always run such simple operations. Therefore we'd rather pass **all** inputs to the elementwise functor as arguments.

SamiAario-AMD replied:
This is now addressed in commit 8c5d7141.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,142 @@
+
+        // Move windows to next block
+        move_tile_windows(x_windows);
+        move_tile_window(y_window, {S::kBlockM});
```
This is not needed if each of your workgroups is processing just single tile.

SamiAario-AMD replied:
Removed in commit 876b057f.

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_operators.hpp:
This is the place for unary elementwise ops: https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/ops/elementwise/unary_element_wise_operation.hpp
You may want to add eg `binary_elementwise_operation.hpp`

SamiAario-AMD replied:
Added in commit fc1ff7fc.

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_shape.hpp:
You might want to reuse this one: https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/ops/common/generic_2d_block_shape.hpp

SamiAario-AMD replied:
As per our discussion on Teams, we will keep using the existing `ElementWiseShape` because it is slightly more appropriate for our implementation.

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_pipeline_default_policy.hpp:
```diff
@@ -0,0 +1,31 @@
+
+#include "ck_tile/core.hpp"
+#include "ck_tile/ops/norm_reduce/block/block_norm_reduce_problem.hpp"
+#include "ck_tile/ops/norm_reduce/block/block_norm_reduce.hpp"
```
These are not used here.

SamiAario-AMD replied:
Removed in commit abd0d356.

aosewski reviewed (changes requested):
Good start! Please add unit-tests (functional) here: https://github.com/ROCm/composable_kernel/tree/develop/test/ck_tile

SamiAario-AMD committed:
Move examples to folder 19_elementwise

SamiAario-AMD committed:
Add missing copyright headers and fix some existing ones

SamiAario-AMD committed:
Replace an assert with throw std::runtime_error in elementwise example

SamiAario-AMD committed:
Avoid reading the output by using make_static_distributed_tensor for y_tile

SamiAario-AMD committed:
Removed two unused includes

SamiAario-AMD committed:
No need to move windows to the next block when each workgroup processes a single tile

SamiAario-AMD committed:
Only copy input tensors to the device

SamiAario-AMD committed:
Use get_warp_size to obtain warp size, and use ceiling division for grid size also for the unary example

Commit:
Adding output strides to the kernel, transposition example and update the other examples

SamiAario-AMD committed:
Changes made by remod.py

SamiAario-AMD committed:
Use default template parameter values for memory operation and coherence in a call to make_naive_tensor_view

SamiAario-AMD committed:
Move binary operations to include/ck_tile/ops/elementwise/binary_elementwise_operation.hpp

SamiAario-AMD committed:
Fix comments in elementwise_example.cpp

- Refer to AMD terminology except when suggesting NVIDIA alternatives in parentheses
- ElementWiseTraits was renamed to ElementWiseShape
- Adopt suggestions made by Copilot when prompted to check for factual or typographical errors

Commit:
Reuse generic reference binary/unary operation in examples + refactoring the transpose reference

SamiAario-AMD committed:
Simplify CMakeLists.txt and remove the unused variables this uncovers

aosewski commented on include/ck_tile/host/reference/reference_transpose.hpp:
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
```suggestion
// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```

SamiAario-AMD replied:
Fixed in 99eac8e2.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
Please rename to `elementwise_kernel.hpp`

SamiAario-AMD replied:
Fixed in 99eac8e2.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise.hpp:
```diff
@@ -0,0 +1,111 @@
+        auto make_tile_windows = [&](const auto& tensors, const auto& iM_) {
+            return generate_tuple(
+                [&](auto idx) {
+                    auto tensor_view = make_naive_tensor_view<address_space_enum::global>(
```
I think we could configure here cache coherence as well. The data won't be reused so we could set up this. 
https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/core/tensor/tensor_view.hpp#L433

SamiAario-AMD replied:
Now I am not sure anymore what you meant by this: https://github.com/ROCm/composable_kernel/pull/2246#discussion_r2113634411

We switched this call to `make_naive_tensor_view` to use the default template parameters based on that note. In any case, now we have two calls to `make_naive_tensor_view` in our kernel with different template parameters, because I neglected to change the other one on line 67. The two calls should still use the same parameters, am I right?

Please confirm if you want the call on line 42 to be switched to:
```
                    auto tensor_view = make_naive_tensor_view<address_space_enum::global,
                                                              memory_operation_enum::set,
                                                              amd_buffer_coherence_enum::slc>(
```

SamiAario-AMD replied:
I did some benchmarking on MI250 and MI300, and the cache coherence settings don't seem to have a noticeable effect on the time it takes to run the kernel. Correctness is not affected.

SamiAario-AMD replied:
As we discussed on Teams, I modified the coherence values on CK tile to have these options for gfx942:
```
enum struct amd_buffer_coherence_enum
{
#if 0
    coherence_default = 0, // default value
    glc               = 1,
    slc               = 2,
    glc_slc           = 3,
#else
    coherence_default = 0, // default value
    WAVE_NT0   = 0,
    WAVE_NT1   = 2,
    GROUP_NT0  = 1,
    GROUP_NT1  = 3,
    DEVICE_NT0 = 8,
    DEVICE_NT1 = 10,
    SYSTEM_NT0 = 9,
    SYSTEM_NT1 = 11,
#endif
};
```
I tested with the default setting and `DEVICE_NT0` for our three examples, but there was no significant performance difference.

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_pipeline_default_policy.hpp:
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
Please fix the copyright year.

SamiAario-AMD replied:
Fixed in 99eac8e2.

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_shape.hpp:
```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright year.

SamiAario-AMD replied:
Fixed in 99eac8e2.

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_pipeline_problem.hpp:
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright year

SamiAario-AMD replied:
Fixed in 99eac8e2.

aosewski commented on include/ck_tile/core/container/tuple.hpp:
```diff
@@ -264,10 +264,14 @@ struct tuple : impl::tuple_base<make_index_sequence<sizeof...(T)>, T...>
-    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get(number<I>) const { TP_COM_(); return get<I>(); }
-    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get()                      { TP_COM_(); return impl::getv<I>(*this); }
-    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get(number<I>)             { TP_COM_(); return get<I>(); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get() const &          { TP_COM_(); return impl::getv<I>(*this); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get(number<I>) const & { TP_COM_(); return get<I>(); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get() &                { TP_COM_(); return impl::getv<I>(*this); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get(number<I>) &       { TP_COM_(); return get<I>(); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get() &&               { TP_COM_(); return impl::getv<I>(std::move(*this)); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get(number<I>) &&      { TP_COM_(); return std::move(*this).template get<I>(); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get() const &&         { TP_COM_(); return impl::getv<I>(std::move(*this)); }
+    template<index_t I> CK_TILE_HOST_DEVICE constexpr decltype(auto) get(number<I>) const &&{ TP_COM_(); return std::move(*this).template get<I>(); }
```
Would that impact performance or register usage?

damien-lejeune replied:
Good question! We haven’t benchmarked this precisely but from what I understand, I don’t think we should be expect any performance or register usage degradation. The changes "const member function" and the lvalue ref-qualifier will have the compiler mostly generate the same code as before: the `impl::getv` already handles returning "T&" and "const T&" and have the same implementation. The change with respect to the rvalue reference, I see three cases from a runtime perspective:
- The object undergoing a move does not have a move constructor (or move operator), the default copy method is used. This is the same as before
- The  object undergoing a move does have a move constructor, or move operator:
    - The implementation of this move constructor is more optimal than the default copy method. This case is improving the perf
    - The implementation of the move constructors is less optimal than the default copy method. This case degrades the performance, but given the nature of the move implementation I think it’s pretty hard to end up in this case.
   
For the use of `std::move`, in principle it should not add runtime overhead as it merely does a casting. So given this I would say these changes would not significantly make the situation worse as before from a runtime performance or register usage perspective. However, I guess `std::move` can potentially impact compile time though.

I'm not sure if this covers what you had in mind. What do you think?

SamiAario-AMD replied:
I'm a bit late to the party, but I added the rvalue implementations because they were required by the `apply` function I added in this branch, so adding the rvalue implementations does not change the performance of existing code.

aosewski reviewed (changes requested):
In general looks good. I got mostly cosmetic changes.

SamiAario-AMD committed:
Rename a file and fix some copyright statements

SamiAario-AMD committed:
Changes made by script/clang-format-overwrite.sh

Commit:
Add basic unit test for ElementWiseKernel

SamiAario-AMD committed:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

SamiAario-AMD committed:
fixup! Use default template parameter values for memory operation and coherence in a call to make_naive_tensor_view

Commit:
Remove left-over uninformative comment in apply unit test

SamiAario-AMD committed:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

SamiAario-AMD committed:
Changes made by clang-format-overwrite.sh

SamiAario-AMD committed:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

aosewski commented on test/ck_tile/container/test_tuple_apply.cpp:
❤️  Really great you've added those tests!

aosewski commented on test/ck_tile/container/test_tuple_apply.cpp:
```diff
@@ -0,0 +1,223 @@
+{
+    // Test with ck_tile::number types
+    auto t      = make_tuple(number<1>{}, number<2>{}, number<3>{});
+    auto result = apply([](auto a, auto b, auto c) { return a.value + b.value + c.value; }, t);
```
number have arithmetic operators overloaded.
```suggestion
    auto result = apply([](auto a, auto b, auto c) { return a + b + c; }, t);
```

SamiAario-AMD replied:
Addressed in commit 186aa4ef.

aosewski commented on test/ck_tile/container/test_tuple_apply.cpp:
```diff
@@ -0,0 +1,223 @@
+
+    // Expected sum: 1 + 2 + ... + N = N*(N+1)/2
+    int expected = N * (N + 1) / 2;
+    EXPECT_EQ(result, expected);
```
I think you can even `static_assert...` that.

SamiAario-AMD replied:
Addressed in commit 186aa4ef.

aosewski commented on test/ck_tile/elementwise/test_elementwise_1d.cpp:
```diff
@@ -0,0 +1,218 @@
+            else
+            {
+                EXPECT_EQ(h_y(i), h_y_ref(i)) << "Error at index " << i;
+            }
```
We usually leverage `check_err` function for checking result correctness along with `compute_atol_rtol` to compute error threshold.

SamiAario-AMD replied:
Addressed in commit 186aa4ef.

aosewski commented on test/ck_tile/elementwise/test_elementwise_1d.cpp:
```diff
@@ -0,0 +1,218 @@
+                                       Shape1_WarpTile,
+                                       Shape1_Vector>;
+
+using TestConfig_HF_Add = std::tuple<ck_tile::half_t,
```
```suggestion
using TestConfig_F16_Add = std::tuple<ck_tile::half_t,
```

SamiAario-AMD replied:
Addressed in commit 186aa4ef.

aosewski commented on test/ck_tile/elementwise/test_elementwise_1d.cpp:
```diff
@@ -0,0 +1,218 @@
+            for(ck_tile::index_t j = 0; j < total_m_elements; ++j)
+            {
+                h_xs[i](j) = static_cast<XDataType>(((i + 1) * (j % 100)) / 10.0f - 5.0f);
+            }
```
We have utility like `FillUniformDistribution` https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/host/fill.hpp

SamiAario-AMD replied:
Addressed in commit 186aa4ef.

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_shape.hpp:
```diff
@@ -0,0 +1,31 @@
+
+    static constexpr index_t kWarpM = WarpTile::at(number<0>{});
+
+    static constexpr index_t kVectorM = Vector::at(number<0>{});
```
(might be in next PR) we'd prefer to infer this value in elementwise pipeline policy. We want it to be maximum possible.

damien-lejeune replied:
Should be addressed in [3a9295d](https://github.com/ROCm/composable_kernel/pull/2246/commits/3a9295d17a5f863aa2936c15f027ae5e383d2ca7)

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -0,0 +1,109 @@
+            return generate_tuple(
+                [&](auto idx) {
+                    auto tensor_view = make_naive_tensor_view<address_space_enum::global>(
+                        tensors.get(idx), lens, input_strides, number<S::kVectorM>{});
```
(might be in next PR) we'd prefer to infer this value in elementwise pipeline policy. We want it to be maximum possible.

aosewski commented on test/ck_tile/elementwise/test_elementwise_1d.cpp:
```diff
@@ -0,0 +1,218 @@
+        h_y_ref.SetZero();
+
+        // Device Buffers
+        std::vector<std::unique_ptr<ck_tile::DeviceMem>> d_xs_mems_owner;
```
You don't have to use `unique_ptr` - the `DeviceMem` object has destructor which frees allocated memory - so it would be called during std::vector destruction.

SamiAario-AMD replied:
It is possible to remove the use of `std::unique_ptr` and `std::make_unique`; this however makes it necessary for `DeviceMem` to have a copy constructor, because `std::vector` apparently requires this. I can add a copy constructor for `DeviceMem` if you want.

SamiAario-AMD replied:
I instead opted to use `std::array` and a utility function `make_uniform_array_with_factory`, to bypass the need to use a copy constructor or copy assignment operator, in commit affa774e.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -0,0 +1,109 @@
+namespace ck_tile {
+
+template <typename Problem_, typename Policy_>
+struct ElementWiseKernel
```
(might be in next PR) - We might want to have the kernel class to calculate the grid size.

aosewski commented on test/ck_tile/elementwise/test_elementwise_1d.cpp:
```diff
@@ -0,0 +1,218 @@
+        ck_tile::ElementWiseKernel<Problem, Policy> ew_kernel;
+
+        // Launch configuration
+        ck_tile::index_t grid_size =
```
(might be in next PR) - We might want to have the kernel class to calculate the grid size.

aosewski commented:
In general looks great! Especially those tests! There are few things which might be addressed in next PR, so this would not wait too long.

SamiAario-AMD committed:
Clean up test_tuple_apply.cpp and test_elementwise_1d.cpp

SamiAario-AMD committed:
Use make_uniform_array_with_factory to define h_xs and d_xs_mems_owner as type std::array

SamiAario-AMD committed:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

SamiAario-AMD committed:
Use a DeviceMem constructor that calls get_element_space_size_in_bytes internally

SamiAario-AMD committed:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

SamiAario-AMD committed:
Move examples to folder 20_elementwise

Commit:
Reduced register pressure on the CK tile elementwise kernel + add 4d input example to be able benchmark against old CK

Commit:
Fix CLang formating

Commit:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

Commit:
Bump up the elementwise example folder number

Commit:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

aosewski commented on example/ck_tile/21_elementwise/elementwise_example_add_4d.cpp:
```diff
@@ -0,0 +1,157 @@
+    auto problem_shape_tuple =
+        ck_tile::make_tuple(problem_shape[0], problem_shape[1], problem_shape[2], problem_shape[3]);
+
+    if(host_strides.size() != 4)
+    {
+        throw std::runtime_error("host_strides must have 4 elements for 4D tensor");
+    }
```
This one is unnecessary cause you have hardcoded `host_stride` vector size.

damien-lejeune replied:
Should be addressed in [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563)

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -30,78 +29,69 @@ struct ElementWiseKernel
+        // Setup block-level coordinates and transforms
+        const index_t iM           = get_block_id() * S::kBlockM;
+        const auto merge_transform = make_merge_transform(
+            ck_tile::generate_tuple([&](auto idx) { return lens[idx]; }, number<lens.size()>{}));
```
If the `lens.size()` size is compile-time known ( and I think it should cause you use tuple for that) you can make a constexpr of this generated tuple.

aosewski replied:
Actually why are you doing this ? Cause it seems to me that you're duplicating the `lens` object ...

damien-lejeune replied:
Good catch! It's indeed copying the lens object ... I think the original intent was to overcome some casting issue when we started develop that part (and had, temporarily a mix of ck and std tuples). It was supposed to be temporary fix but somehow remained there. So thanks for pointing this out.

Addressed in [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563)

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -30,78 +29,69 @@ struct ElementWiseKernel
+        const index_t iM           = get_block_id() * S::kBlockM;
+        const auto merge_transform = make_merge_transform(
+            ck_tile::generate_tuple([&](auto idx) { return lens[idx]; }, number<lens.size()>{}));
+        const auto dim_seq        = make_index_sequence<Dims::size()>{};
+        const auto merged_dim_seq = ck_tile::make_tuple(sequence<0>{});
```
I think you can put this code inline to the `transform_tensor_view` call in line 48. And both of them are constexpr.

damien-lejeune replied:
Addressed in [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563)

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -30,78 +29,69 @@ struct ElementWiseKernel
+        // Setup output window and store the result tile.
+        const auto y_m_n = make_naive_tensor_view<address_space_enum::global,
+                                                  memory_operation_enum::set,
+                                                  amd_buffer_coherence_enum::slc>(
```
This setting is valid only for arch < gfx94

damien-lejeune replied:
Addressed in [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563)

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -30,78 +29,69 @@ struct ElementWiseKernel
-                                  ck_tile::make_tuple(sequence<0>{}));
+        const auto transformed_y_m_n = transform_tensor_view(y_m_n,
+                                                             ck_tile::make_tuple(merge_transform),
+                                                             ck_tile::make_tuple(dim_seq),
+                                                             merged_dim_seq);
```
Ok, you're reusing `dim_seq` and `merged_dim_seq` here... Anyway still those are constexpr which you can put inline here IMHO.

damien-lejeune replied:
Addressed in [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563)

aosewski commented on test/ck_tile/elementwise/test_elementwise_1d.cpp:
```diff
@@ -0,0 +1,209 @@
+
+TYPED_TEST(TestCkTileElementwise, RunElementwise_512)
+{
+    this->RunTest(512); // Test with a size that might not be a multiple of blockM
```
the 512 is a multiple of 256

damien-lejeune replied:
Set to 513. [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563)

aosewski commented on test/ck_tile/elementwise/test_elementwise_1d.cpp:
```diff
@@ -0,0 +1,209 @@
+    this->RunTest(512); // Test with a size that might not be a multiple of blockM
+}
+
+TYPED_TEST(TestCkTileElementwise, RunElementwise_Small_32)
```
Please add a test case for odd size as well.

damien-lejeune replied:
Addressed in [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563)

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -0,0 +1,99 @@
+                    input_tensors.get(i), lens, input_strides, number<S::kVectorM>{});
+
+                const auto transformed_tensor =
+                    transform_tensor_view(tensor_view,
```
I think we need to do padding here as well to support sizes that are not a multiple of block size

damien-lejeune replied:
Addressed in [e6d00bf](https://github.com/ROCm/composable_kernel/pull/2246/commits/e6d00bfee49d6932adeb1e0a8283d6b1a1374563).

I added the padding and tested the performance. It's slightly slower by a couple of percents. I would say it's worth it.

aosewski commented on include/ck_tile/ops/elementwise/kernel/elementwise_kernel.hpp:
```diff
@@ -0,0 +1,99 @@
+        const auto x_tiles = ck_tile::generate_tuple(
+            [&](auto i) {
+                const auto tensor_view = make_naive_tensor_view<address_space_enum::global>(
+                    input_tensors.get(i), lens, input_strides, number<S::kVectorM>{});
```
We might want to have an `IsSupportedArg` function which would see if we can support specific test case - eg if rightmost dimension is a multiple of vector size.

damien-lejeune replied:
Does it still matter if we use padding?

aosewski replied:
Yes, since if we use vectorized reads the input has to be a multiple of vector size, otherwise we would get errors.

damien-lejeune replied:
Addressed in [612a8c3](https://github.com/ROCm/composable_kernel/pull/2246/commits/612a8c3b1bdb4deb33157f4f5e277a4e7c49fb96)

Commit:
Elementwise: add padding + minor cleanup

Commit:
Add Vector Size inference + fix issue with wrong vectorization due to missing GuaranteedLastDimensionVectorStride setting in make_naive_tensor_view

aosewski commented on include/ck_tile/ops/elementwise/pipeline/elementwise_shape.hpp:
```diff
@@ -7,25 +7,24 @@
 
-    static constexpr index_t kVectorM = Vector::at(number<0>{});
+    static constexpr index_t kVectorM =
+        16 / sizeof(ComputeDataType); // There are 16 32-bits lanes per SIMD
```
This comment is confusing.

damien-lejeune replied:
Addressed in [612a8c3](https://github.com/ROCm/composable_kernel/pull/2246/commits/612a8c3b1bdb4deb33157f4f5e277a4e7c49fb96)

Commit:
Merge branch 'develop' into LWPCK-3170-elementwise-kernel-general

Commit:
Add isSupportedArg to Elementwise kernel + addapt example and unit tests

Commit:
Fix clang-format on the unit test file
